weasyl/searchtag.py

```diff
@@ -217,3 +253,163 @@ def tag_history(submitid):
+    tags as appropriate
+
+    Parameters:
+        userid: The userid of the user submitting the request.
+        tags: A set() object of tags; must have been passed through ``parse_blacklist_tags()``
+                (occurs in the the controllers/settings.py controller)
+        edit_global_blacklist: Optional. Set to Boolean True if the global blacklist is to be edited.
+
+    Returns:
+        Nothing.
+    """
+    # Only directors can edit the global blacklist; sanity check against the @director_only decorator
+    if edit_global_blacklist and userid not in staff.DIRECTORS:
+        raise WeasylError("InsufficientPermissions")
+
+    # Determine what, if any, tags exist before editing, depending on what we are editing
+    if not edit_global_blacklist:
+        existing = d.engine.execute("""
+            SELECT tagid FROM searchmapuserblacklist WHERE userid = %(uid)s
+        """, uid=userid).fetchall()
+    else:
+        existing = d.engine.execute("""
+            SELECT tagid FROM searchmapglobalblacklist
+        """).fetchall()
+
+    # Get the tag titles/ids out of the searchtag table
+    query = d.engine.execute("""
+        SELECT tagid, title FROM searchtag WHERE title = ANY (%(title)s)
+    """, title=list(tags)).fetchall()
+
+    # Determine which (if any) of the valid tags are new; add them to the searchtag table if so.
+    newtags = list(tags - {x.title for x in query})
+    if newtags:
+        query.extend(
+            d.engine.execute(
+                "INSERT INTO searchtag (title) SELECT * FROM UNNEST (%(newtags)s) AS title RETURNING tagid, title",
+                newtags=newtags
+            ).fetchall())
+
+    existing_tagids = {t.tagid for t in existing}
+    entered_tagids = {t.tagid for t in query}
+
+    # Assign added and removed
+    added = entered_tagids - existing_tagids
+    removed = existing_tagids - entered_tagids
+
+    if added:
+        if edit_global_blacklist:
+            d.engine.execute("""
+                INSERT INTO searchmapglobalblacklist
+                    SELECT tag, %(uid)s
+                    FROM UNNEST (%(added)s) AS tag
+            """, uid=userid, added=list(added))
+        else:
+            d.engine.execute("""
+                INSERT INTO searchmapuserblacklist
+                SELECT tag, %(uid)s
+                FROM UNNEST (%(added)s) AS tag
+            """, uid=userid, added=list(added))
+
+    if removed:
+        if edit_global_blacklist:
+            d.engine.execute("""
+                DELETE FROM searchmapglobalblacklist
+                WHERE tagid = ANY (%(removed)s)
+            """, removed=list(removed))
+        else:
+            d.engine.execute("""
+                DELETE FROM searchmapuserblacklist
+                WHERE userid = %(uid)s AND tagid = ANY (%(removed)s)
+            """, uid=userid, removed=list(removed))
+
+
+def get_searchtag_blacklist(userid, global_blacklist=False):
```

Comment:
Write separate functions to get the global and user-specific blacklists. The two paths have nothing in common in this function.